### PR TITLE
Group Normalization

### PR DESCRIPTION
**Pull request by ssnl**
Implements group normalization.

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -1665,7 +1665,7 @@ def test_InstanceNorm3d_general_cuda(self):
     def _test_LayerNorm_general(self, type):
         for i in range(2, 6):
             shape = torch.LongTensor(i).random_(3, 6).tolist()
-            x = Variable(type(*shape).uniform_(0, 10))
+            x = type(*shape).uniform_(0, 10)
```
Comment (SsnL) [inline test/test_nn.py]: remove Variable buffer

File: test/test_nn.py [hunk 2]
```diff
@@ -1712,8 +1712,7 @@ def _test_LayerNorm_general(self, type):
             self.assertEqual(old_running_var, ln.running_var)
 
     def _test_LayerNorm_cuda_half(self):
-        # just THNN, LayerNorm has no cuDNN path
-        input = Variable(torch.rand(2, 3, 3, 2).cuda().half().random_(1, 10), requires_grad=True)
+        input = torch.zeros(2, 3, 3, 2, requires_grad=True).cuda().half().random_(1, 10)
```
Comment (SsnL) [inline test/test_nn.py]: remove Variable buffer

File: test/test_nn.py [hunk 3]
```diff
@@ -1727,6 +1726,93 @@ def test_LayerNorm_general_cuda(self):
...
+            gn.weight.data.copy_(scale)
+            gn.bias.data.copy_(bias)
+            output = gn(x)
+            out_reshaped = output.view(b, c, -1)
+            out_normed = (out_reshaped - bias.view(c, 1)) / scale.view(c, 1)
+            out_normed_reshaped = out_normed.view(b, g, -1)
+            mean = out_normed_reshaped.mean(-1)
+            var = out_normed_reshaped.var(-1, unbiased=False)
+            self.assertAlmostEqual(torch.abs(mean).mean(), 0, delta=1e-5)
+            self.assertAlmostEqual(torch.abs(var).mean(), 1, delta=1e-5)
```
Comment (SsnL) [inline test/test_nn.py]: tests that the data has unit var and 0 mean after undoing the affine transform

File: test/test_nn.py [hunk 4]
```diff
@@ -5759,6 +5845,70 @@ def multimarginloss_weights_no_reduce_test():
...
+        desc='2d_no_affine_LN',
+    ),
+    dict(
+        module_name='GroupNorm',
+        constructor_args=(2, 4, 1e-3, 0.3, True, True),
+        input_size=(4, 4, 2, 3),
+        cudnn=True,
+        check_eval=True,
+        desc='2d_affine_tracking_stats',
+    ),
```
Comment (SsnL) [inline test/test_nn.py]: Testing higher dimensional cases is not necessary as it goes through the same `.view(1, batch * group, -1)` reshape.

File: test/test_nn.py [hunk 5]
```diff
@@ -5759,6 +5845,70 @@ def multimarginloss_weights_no_reduce_test():
...
+        check_eval=True,
+        desc='2d_affine',
+    ),
+    dict(
+        module_name='GroupNorm',
+        constructor_args=(3, 3, 1e-3, 0.3, False),
+        input_size=(4, 3, 2, 3),
+        cudnn=True,
+        check_eval=True,
+        desc='2d_no_affine_IN',
```
Comment (SsnL) [inline test/test_nn.py]: This test case is using GN to simulate an IN

File: test/test_nn.py [hunk 6]
```diff
@@ -5759,6 +5845,70 @@ def multimarginloss_weights_no_reduce_test():
...
+    ),
+    dict(
+        module_name='GroupNorm',
+        constructor_args=(1, 3, 1e-3, 0.3, False),
+        input_size=(4, 3, 2, 3),
+        cudnn=True,
+        check_eval=True,
+        desc='2d_no_affine_LN',
```
Comment (SsnL) [inline test/test_nn.py]: This test case is using GN to simulate an LN

File: torch/nn/functional.py [hunk 7]
```diff
@@ -1206,8 +1206,8 @@ def batch_norm(input, running_mean, running_var, weight=None, bias=None,
     )
 
 
-def instance_norm(input, running_mean, running_var, weight=None, bias=None,
-                  use_input_stats=True, momentum=0.1, eps=1e-5):
+def instance_norm(input, running_mean=None, running_var=None, weight=None,
```
Comment (SsnL) [inline torch/nn/functional.py]: Running stats should really be optional here as they are not typically used in IN

File: torch/nn/functional.py [hunk 8]
```diff
@@ -1256,7 +1256,7 @@ def _instance_norm(input, running_mean=None, running_var=None, weight=None,
                           eps=eps)
 
 
-def layer_norm(input, normalized_shape, running_mean, running_var,
+def layer_norm(input, normalized_shape, running_mean=None, running_var=None,
```
Comment (SsnL) [inline torch/nn/functional.py]: Running stats should really be optional here as they are not typically used in LN

File: torch/nn/functional.py [hunk 9]
```diff
@@ -1266,6 +1266,16 @@ def layer_norm(input, normalized_shape, running_mean, running_var,
...
+    if weight is not None and weight.size() != normalized_shape:
+        raise ValueError('Expected weight to be of same shape as '
+                          'normalized_shape, but got {} weight and '
+                          'normalized_shape={}'.format(weight.size(), normalized_shape))
+
+    if bias is not None and bias.size() != normalized_shape:
+        raise ValueError('Expected bias to be of same shape as '
+                          'normalized_shape, but got {} bias and '
+                          'normalized_shape={}'.format(bias.size(), normalized_shape))
+
```
Comment (SsnL) [inline torch/nn/functional.py]: Adds more shape checking

File: torch/nn/functional.py [hunk 10]
```diff
@@ -1308,6 +1318,73 @@ def layer_norm(input, normalized_shape, running_mean, running_var,
...
+    if running_mean is not None:
+        running_mean_orig = running_mean
+        running_mean = running_mean_orig.repeat(b)
+    if running_var is not None:
+        running_var_orig = running_var
+        running_var = running_var_orig.repeat(b)
+
+    # Apply group norm
+    input_reshaped = input.contiguous().view(1, b * g, -1)
+
```
Comment (SsnL) [inline torch/nn/functional.py]: Basically we make BN kernels to normalize the `[c // g, *]` part of `[b, g, c // g, *]`

**Comment by SsnL** [comment]
cc @kaiminghe

File: torch/nn/functional.py [hunk 11]
```diff
@@ -1308,6 +1318,53 @@ def layer_norm(input, normalized_shape, running_mean, running_var,
...
+                         'number of channels in input, but got {} weight and {} '
+                         'input'.format(weight.size(), input_shape))
+
+    if bias is not None and (bias.dim() != 1 or bias.numel() != c):
+        raise ValueError('Expected bias to be a vector of size equal to the '
+                         'number of channels in input, but got {} bias and {} '
+                         'input'.format(bias.size(), input_shape))
+
+    # Apply group norm
+    input_reshaped = input.contiguous().view(1, b * g, -1)
```
Comment (fmassa) [inline torch/nn/functional.py]: nit: this can be simplified with `.reshape(1, b * g, -1)`, which was added in https://github.com/pytorch/pytorch/pull/5575 and consists roughly as `.contiguous().view()`

**Comment by SsnL** [comment]
Right. However, here we actually want it to be contiguous to take advantage
of the cudnn path. Reshape doesn’t always make it contiguous.

File: torch/nn/functional.py [hunk 12]
```diff
@@ -1308,6 +1318,53 @@ def layer_norm(input, normalized_shape, running_mean, running_var,
         return out
 
 
+def group_norm(input, num_groups, weight=None, bias=None, eps=1e-5):
```
Comment (ezyang) [inline torch/nn/functional.py]: Why not in ATen?

**Comment by SsnL** [comment]
Fair. Let me make it in ATen.

**Comment by ezyang** [comment]
Looks good, but I did not do a math check. @KaimingHe, maybe you can look at it?

**Comment by SsnL** [comment]
@ezyang Moved the code to ATen :). Now one of the tests runs 2x as fast. I should move LN and IN to Aten as well.

**Comment by KaimingHe** [comment]
Looks good for me.

**Comment by ppwwyyxx** [comment]
BTW I was able to reproduce our ImageNet experiments in Pytorch.
I used fewer augmentations (those in official pytorch examples) than our original experiments, and got 24.1 and 24.2 val error in two runs. The number reported in the paper is 24.0~24.2.

**Comment by soumith** [comment]
@ppwwyyxx that's awesome!
